plugins/woocommerce/src/Internal/Traits/RestApiCache.php

```diff
@@ -0,0 +1,405 @@
+	 */
+	private function build_cache_config( WP_REST_Request $request, array $config ): ?array {
+		$entity_type = $config['entity_type'] ?? $this->get_default_entity_type();
+
+		if ( ! $entity_type ) {
+			wc_get_container()->get( LegacyProxy::class )->call_function(
+				'wc_doing_it_wrong',
+				__METHOD__,
+				'No entity type provided and no default entity type available. Skipping cache.',
+				'10.4.0'
+			);
+			return null;
+		}
+
+		return array(
+			'entity_type' => $entity_type,
+			'cache_key'   => $this->get_cache_key( $request, $entity_type ),
+		);
+	}
+
+	/**
+	 * Cache the response if it's successful and return it with appropriate headers.
+	 *
+	 * Only caches responses with 2xx status codes. Always adds the X-WC-Cache header
+	 * with value MISS if the response was cached, or SKIP if it was not cached.
+	 *
+	 * Supports both WP_REST_Response objects and raw data (which will be wrapped in a response object).
+	 * Error objects are returned as-is without caching.
+	 *
+	 * @param WP_REST_Request                         $request       The request object.
+	 * @param WP_REST_Response|\WP_Error|array|object $response      The response to potentially cache.
+	 * @param array                                   $cached_config Caching configuration from build_cache_config().
+	 * @return WP_REST_Response|\WP_Error The response with appropriate cache headers.
+	 */
+	private function maybe_cache_response( WP_REST_Request $request, $response, array $cached_config ) {
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response = rest_ensure_response( $response );
+
+		$cached = false;
+
+		$status = $response->get_status();
+		if ( $status >= 200 && $status <= 299 ) {
+			$data       = $response->get_data();
+			$entity_ids = is_array( $data ) ? $this->extract_entity_ids( $data ) : array();
+
+			$this->store_cached_response(
+				$cached_config['cache_key'],
+				$data,
+				$status,
+				$cached_config['entity_type'],
+				$entity_ids
+			);
+
+			$cached = true;
+		}
+
+		$response->header( 'X-WC-Cache', $cached ? 'MISS' : 'SKIP' );
+		return $response;
+	}
+
+	/**
+	 * Get the default type for entities included in responses.
+	 *
+	 * This can be customized per-endpoint via the config array
+	 * passed to with_cache() ('entity_type' key).
+	 *
+	 * @return string|null Entity type (e.g., 'product', 'order'), or null if no controller-wide default.
+	 */
+	protected function get_default_entity_type(): ?string {
+		return null;
+	}
+
+	/**
+	 * Extract entity IDs from response data.
+	 *
+	 * This implementation assumes the response is either:
+	 * - An array with an 'id' field (single item)
+	 * - An array of arrays each having an 'id' field (collection)
+	 *
+	 * @param array $data Response data.
+	 * @return array Array of entity IDs.
+	 */
+	private function extract_entity_ids( array $data ): array {
```

Comment:
I'm not quite following the need for extracting IDs.  Is the idea behind this function that we can optimistically set the caches for individual items in addition to the current response?

If so, I think there may be some risk here with provided query params and differences in endpoints.